macaroon.js

```diff
@@ -338,197 +392,249 @@ const readFieldV2Optional = function(buf, maybeFieldType) {
-      _vid: null,
-      _location: null,
-    };
-    if (verificationId) {
-      caveat._location = requireString(location, 'Macaroon caveat location');
-      caveat._vid = bytesToBitArray(
-        requireUint8Array(verificationId, 'Macaroon caveat verification id'));
-    }
-    this._caveats.push(caveat);
-    this._signature = keyedHash2(
-      this._signature, caveat._vid, sjcl.codec.utf8String.toBits(caveatId));
+   * Return the signature of the macaroon.
+   * @return {string} The macaroon's signature.
+   */
+  get signature() {
+    return bitsToBytes(this._signatureBits);
   }
 
   /**
     Adds a third party caveat to the macaroon. Using the given shared root key,
     caveat id and location hint. The caveat id should encode the root key in
     some way, either by encrypting it with a key known to the third party or by
     holding a reference to it stored in the third party's storage.
-    @param {bitArray} rootKey
-    @param {String} caveatId
-    @param {String} location
+    @param {Uint8Array} rootKeyBytes
+    @param {Uint8Array | string} caveatIdBytes
+    @param {optional String} locationStr
   */
-  addThirdPartyCaveat(rootKey, caveatId, location) {
-    const verificationId = bitArrayToUint8Array(
-      encrypt(this._signature,
-        makeKey(requireUint8Array(rootKey, 'Caveat root key'))));
-    this.addCaveat(
-      requireString(caveatId, 'Caveat id'),
-      verificationId,
-      requireString(location, 'Caveat location'));
+  addThirdPartyCaveat(rootKeyBytes, caveatIdBytes, locationStr) {
+    const cav = {
+      _identifierBits: bytesToBits(requireBytes(caveatIdBytes, 'Caveat id')),
+      _vidBits: encrypt(
+        this._signatureBits,
+        makeKey(bytesToBits(requireBytes(rootKeyBytes, 'Caveat root key')))),
+      _locationStr: maybeString(locationStr),
+    };
+    this._signatureBits = keyedHash2(
+      this._signatureBits,
+      cav._vidBits,
+      cav._identifierBits
+    );
+    this._caveats.push(cav);
   }
 
   /**
     Adds a caveat that will be verified by the target service.
-    @param {String} caveatId
+    @param {String | Uint8Array} caveatId
   */
-  addFirstPartyCaveat(caveatId) {
-    this.addCaveat(caveatId, null, null);
+  addFirstPartyCaveat(caveatIdBytes) {
+    const identifierBits = bytesToBits(requireBytes(caveatIdBytes, 'Condition'));
+    this._caveats.push({
+      _identifierBits: identifierBits,
+    });
+    this._signatureBits = keyedHash(this._signatureBits, identifierBits);
   }
 
   /**
-    Sets the macaroon signature to one bound to the given signature.
+    Sets the macaroon signature to one bound to the given root signature.
     This must be called on discharge macaroons with the primary
     macaroon's signature before sending the macaroons in a request.
-    @param {Uint8Array} sig
+    @param {Uint8Array} rootSig
   */
-  bind(sig) {
-    sig = bytesToBitArray(sig);
-    this._signature = bindForRequest(sig, this._signature);
+  bind(rootSig) {
```

Comment:
I think it's probably fine, but `bind` is a built-in Function method so might be some confusion.

Reply:
renamed to bindToRoot.